Edits/Updates to API documentation...

chrislkeller opened the pull request:
Hey...

I was working through making requests to the DocumentCloud API today and found a couple of discrepancies when it came to passing authentication keys through a header. The [docs say to preface](https://www.documentcloud.org/help/api#authentication) with `Bearer`, but that didn't work. Once I changed it to `Token` — which matches the [MuckRock docs](https://www.muckrock.com/api/) — it worked.

mitchelljkotler commented:
Hi Chris,

Thanks for helping out with our documentation.  Would you be able to provide the API calls you were having trouble with and the ones that worked for you?  Were you using the MuckRock API?  The MuckRock API does use Token in the header, but the documentation for DocumentCloud is correct - it does expect Bearer.  Also, the API token in your MuckRock profile, which you reference, is only valid for MuckRock API calls and not for DocumentCloud API calls, so I am somewhat confused as to the issue you are facing.

Thanks,
Mitch

chrislkeller commented:
Hi Mitch...

Thank you for the response. I'm super confused as to what happened to allow things to work and led to my edits, so let's walk through where I'm at.

I am using the API token found on my MuckRock account profile. In reading what is [outlined here](https://www.documentcloud.org/help/api#authentication), I understood to be the approach.

  >Authentication happens at the MuckRock accounts server located at https://accounts.muckrock.com/. The API provided there will supply you with a JWT access token and refresh token in exchange for your username and password.

I'm not sure where the "API provided there" is because the docs say `Authentication happens at the MuckRock accounts server located at https://accounts.muckrock.com`

And at my level, it's not entirely clear how to make an authenticated call to the DocumentCloud API. This line at the top of the docs — `All APIs besides the authentication endpoints are served from https://api.www.documentcloud.org/api.` — makes me think I should I be using a different endpoint. 

I guess I'm left to assume the proper way is to POST to `https://api.www.documentcloud.org/api/token/` which will yield the proper token?

Perhaps this is where I stumbled.

All I know is I was limited to 100 documents in a response, but when I added my MuckRock token to the requests headers and changed `Bearer` to `Token` I was able to retrive more than 8,000 documents in a request.

Shrug?

Let me go through it step by step.

**Use case**: Get the title, slug, page count and other data for all the public documents are associated with our organization's account which I believe requires authentication?

**Headers I'm passing when it fails**

  ```
  REQUEST_HEADERS = {
      "From": "email@email.email",
      "Reason": "Research for news stories",
      "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/58.0.3029.110 Safari/537.3",
      "Authorization": "Bearer %s" % _DC_TOKEN,
  }
  ```

**The request**

  ```
  url = https://api.www.documentcloud.org/api/documents/?access=public&organization=<our-id>&format=json&per_page=100

  def make_request(url):
      req = requests.get(url, headers=REQUEST_HEADERS)
      if req.status_code == requests.codes.ok:
          response = req.json()
          logger.info("Found %s documents" % len(response["results"]))
          logger.info(response)
          return response
      else:
          logger.info(req.status_code)
          return False
  ```

**Result**

  ```
  pipenv run python 1-inventory-ap-public-documents.py
  (def make_request 140):  403
  (def _init_ 103):  Request failed
  ```
----

**Headers I'm passing when it works**

  ```
  REQUEST_HEADERS = {
      "From": "email@email.email",
      "Reason": "Research for news stories",
      "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/58.0.3029.110 Safari/537.3",
      "Authorization": "Token %s" % _DC_TOKEN,
  }
  ```
**The request**

  ```
  url = https://api.www.documentcloud.org/api/documents/?access=public&organization=<our-id>&format=json&per_page=100
  def make_request(url):
      req = requests.get(url, headers=REQUEST_HEADERS)
      if req.status_code == requests.codes.ok:
          response = req.json()
          logger.info("Found %s documents" % len(response["results"]))
          logger.info(response)
          return response
      else:
          logger.info(req.status_code)
          return False
  ```
**Result**

  ```
  pipenv run python 1-inventory-ap-public-documents.py
  (def _init_ 86):  https://api.www.documentcloud.org/api/documents/?access=public&cursor=cD00MTQyMDAy&format=json&organization=88
  (def make_request 136):  Found 25 documents
  (def make_request 137):  {'next': 'https://api.www.documentcloud.org/api/documents/?access=public&cursor=cD00MTQyMDQw&format=json&organization=88', 'previous': 'https://api.www.documentcloud.org/api/documents/?access=public&cursor=cj0xJnA9NDE0MjAwMw%3D%3D&format=json&organization=88', 'results': [{'id': 4142003, 'access': 'public', 'asset_url': 'https://s3.documentcloud.org/'...
  ```

I think it's safe to spike this pull request. Somewhere, things got all mixed up and I'm likely the culprit of mixing and matching things. Or if there's something I am missing and you can help a journalist programmer fill in the gaps I'm more than happy to edit the changes I made to spell out how a beginning coder can get authenticated.

Chris K.

mitchelljkotler commented:
I see you are using Python, so by far the easiest thing for you to do would
be to use our Python library here:
https://documentcloud.readthedocs.io/en/latest/
This will allow you to use your username and password and it automatically
handles all of the API calls to get you a valid JWT token and validate your
requests.  It should also make it easier to request the data you need.  I
am happy to help with that if needed.

The way you are doing it now is not properly authenticating your requests,
and they are going through anonymously.

On Fri, Jun 9, 2023 at 12:39 PM Chris Keller ***@***.***>
wrote:

> Hi Mitch...
>
> Thank you for the response. I'm super confused as to what happened to
> allow things to work and led to my edits, so let's walk through where I'm
> at.
>
> I am using the API token found on my MuckRock account profile. In reading
> what is outlined here
> <https://www.documentcloud.org/help/api#authentication>, I understood to
> be the approach.
>
> Authentication happens at the MuckRock accounts server located at
> https://accounts.muckrock.com/. The API provided there will supply you
> with a JWT access token and refresh token in exchange for your username and
> password.
>
> I'm not sure where the "API provided there" is because the docs say Authentication
> happens at the MuckRock accounts server located at
> https://accounts.muckrock.com
>
> And at my level, it's not entirely clear how to make an authenticated call
> to the DocumentCloud API. This line at the top of the docs — All APIs
> besides the authentication endpoints are served from
> https://api.www.documentcloud.org/api. — makes me think I should I be
> using a different endpoint.
>
> I guess I'm left to assume the proper way is to POST to
> https://api.www.documentcloud.org/api/token/ which will yield the proper
> token?
>
> Perhaps this is where I stumbled.
>
> All I know is I was limited to 100 documents in a response, but when I
> added my MuckRock token to the requests headers and changed Bearer to
> Token I was able to retrive more than 8,000 documents in a request.
>
> Shrug?
>
> Let me go through it step by step.
>
> *Use case*: Get the title, slug, page count and other data for all the
> public documents are associated with our organization's account which I
> believe requires authentication?
>
> *Headers I'm passing when it fails*
>
> REQUEST_HEADERS = {
>     "From": ***@***.***",
>     "Reason": "Research for news stories",
>     "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/58.0.3029.110 Safari/537.3",
>     "Authorization": "Bearer %s" % _DC_TOKEN,
> }
>
> *The request*
>
> url = https://api.www.documentcloud.org/api/documents/?access=public&organization=<our-id>&format=json&per_page=100
>
> def make_request(url):
>     req = requests.get(url, headers=REQUEST_HEADERS)
>     if req.status_code == requests.codes.ok:
>         response = req.json()
>         logger.info("Found %s documents" % len(response["results"]))
>         logger.info(response)
>         return response
>     else:
>         logger.info(req.status_code)
>         return False
>
> *Result*
>
> pipenv run python 1-inventory-ap-public-documents.py
> (def make_request 140):  403
> (def _init_ 103):  Request failed
>
> ------------------------------
>
> *Headers I'm passing when it works*
>
> REQUEST_HEADERS = {
>     "From": ***@***.***",
>     "Reason": "Research for news stories",
>     "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/58.0.3029.110 Safari/537.3",
>     "Authorization": "Token %s" % _DC_TOKEN,
> }
>
> *The request*
>
> url = https://api.www.documentcloud.org/api/documents/?access=public&organization=<our-id>&format=json&per_page=100
> def make_request(url):
>     req = requests.get(url, headers=REQUEST_HEADERS)
>     if req.status_code == requests.codes.ok:
>         response = req.json()
>         logger.info("Found %s documents" % len(response["results"]))
>         logger.info(response)
>         return response
>     else:
>         logger.info(req.status_code)
>         return False
>
> *Result*
>
> pipenv run python 1-inventory-ap-public-documents.py
> (def _init_ 86):  https://api.www.documentcloud.org/api/documents/?access=public&cursor=cD00MTQyMDAy&format=json&organization=88
> (def make_request 136):  Found 25 documents
> (def make_request 137):  {'next': 'https://api.www.documentcloud.org/api/documents/?access=public&cursor=cD00MTQyMDQw&format=json&organization=88', 'previous': 'https://api.www.documentcloud.org/api/documents/?access=public&cursor=cj0xJnA9NDE0MjAwMw%3D%3D&format=json&organization=88', 'results': [{'id': 4142003, 'access': 'public', 'asset_url': 'https://s3.documentcloud.org/'...
>
> I think it's safe to spike this pull request. Somewhere, things got all
> mixed up and I'm likely the culprit of mixing and matching things. Or if
> there's something I am missing and you can help a journalist programmer
> fill in the gaps I'm more than happy to edit the changes I made to spell
> out how a beginning coder can get authenticated.
>
> Chris K.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/MuckRock/documentcloud/pull/142#issuecomment-1584857724>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABBHIWDA6CAFEKEMOU473TXKNGTDANCNFSM6AAAAAAY73RTQM>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>
